src/particles/ShapeFactors.H

```diff
@@ -107,83 +107,52 @@ int compute_shape_factor <3> (amrex::Real* const sx, amrex::Real xmid){
+    amrex::Real factor;
+    int cell;
+};
 
-    const auto i = static_cast<int>(amrex::Math::floor(x_old));
-    const int i_shift = i - i_new;
-    const amrex::Real xint = x_old - i;
-    sx[1+i_shift] = 1.0_rt - xint;
-    sx[2+i_shift] = xint;
-    return i;
-}
-
-/**
- *  Compute shifted shape factor and return index of leftmost cell where
- *  particle writes, for Esirkepov algorithm.
- *  Specialization for order 2
- */
-template <>
-AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
-int compute_shifted_shape_factor <2> (amrex::Real* const sx,
-                                      const amrex::Real x_old,
-                                      const int i_new){
-
-    using namespace amrex::literals;
-
-    const auto i = static_cast<int>(amrex::Math::floor(x_old+0.5_rt));
-    const int i_shift = i - (i_new + 1);
-    const amrex::Real xint = x_old - i;
-    sx[1+i_shift] = 0.5_rt*(0.5_rt-xint)*(0.5_rt-xint);
-    sx[2+i_shift] = 0.75_rt-xint*xint;
-    sx[3+i_shift] = 0.5_rt*(0.5_rt+xint)*(0.5_rt+xint);
-    // index of the leftmost cell where particle deposits
-    return i-1;
-}
-
-/**
- *  Compute shifted shape factor and return index of leftmost cell where
- *  particle writes, for Esirkepov algorithm.
- *  Specialization for order 3
+/** \brief Compute a single shape factor and return the index of the cell where the particle writes.
+ *
+ * \tparam depos_order Order of the shape factor
+ * \param[in] xmid exact position of the particle in index space
+ * \param[in] ix index of the shape factor, must be 0 <= ix <= depos_order
  */
-template <>
-AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
-int compute_shifted_shape_factor <3> (amrex::Real* const sx,
-                                      const amrex::Real x_old,
-                                      const int i_new){
+template<int depos_order> AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+shape_factor_result compute_shape_factor_branchless (amrex::Real xmid, int ix) noexcept {
 
     using namespace amrex::literals;
 
-    const auto i = static_cast<int>(amrex::Math::floor(x_old));
-    const int i_shift = i - (i_new + 1);
-    const amrex::Real xint = x_old - static_cast<amrex::Real>(i);
-    sx[1+i_shift] = 1.0_rt/6.0_rt*(1.0_rt-xint)*(1.0_rt-xint)*(1.0_rt-xint);
-    sx[2+i_shift] = 2.0_rt/3.0_rt-xint*xint*(1.0_rt-xint/2.0_rt);
-    sx[3+i_shift] = 2.0_rt/3.0_rt-(1.0_rt-xint)*(1.0_rt-xint)*(1.0_rt-0.5_rt*(1.0_rt-xint));
-    sx[4+i_shift] = 1.0_rt/6.0_rt*xint*xint*xint;
-    // index of the leftmost cell where particle deposits
-    return i-1;
+    if constexpr (depos_order==0) {
+        const int i = static_cast<int>(std::floor(xmid+0.5));
+        return {1, i+ix};
+    } else if constexpr (depos_order==1) {
+        const int i = static_cast<int>(std::floor(xmid));
+        const amrex::Real xint = xmid-i;
+        const amrex::Real ic = static_cast<amrex::Real>(ix);
+        const amrex::Real shape = (1._rt-ic) + (2._rt*ic-1._rt) * xint;
+        return {shape, i+ix};
```

Comment:
`xmid` is cast to `int` and re-cast to `Real` in the next line. Could you do this instead? I believe it applies equally to all `depos_order`. It spares 1 implicit conversion, but I don't know if it'll be better.
```suggestion
        const amrex::Real i_floor = std::floor(xmid);
        const amrex::Real xint = xmid-i_floor;
        const amrex::Real ic = static_cast<amrex::Real>(ix);
        const amrex::Real shape = (1._rt-ic) + (2._rt*ic-1._rt) * xint;
        return {shape, static_cast<int>(i_floor)+ix};
```